### PR TITLE
Add gradient calculation for the covariance between points in GPyModelWrapper

### PR DESCRIPTION
**Pull request by BrunoKM**
This change adds a method `get_covariance_between_points_gradients` to the `GPyModelWrapper` to facilitate obtaining the gradients of the corresponding `get_covariance_between_points` method implemented in the wrapper.

These gradients are useful when optimising any batch acquisition function that relies on the covariance between batch points (such as Multi-point Expected Improvement) sequentially, i.e. when optimising with respect to one element of the batch only.

A second change introduced is vectorization of the gradient calculation for full covariance in the `GPyModelWrapper` class. Replacing the for-loop with numpy array operations results in ~50% speed-up for larger batches of points (> 50).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by apaleyes** [comment]
Apologies for taking time on this one, it's a bit hard for me to make complete sense of this change. I promise to get through it!

**Comment by BrunoKM** [comment]
No worries, let me know if there is anything I can do to explain, document or motivate it better!

File: emukit/model_wrappers/gpy_model_wrappers.py [hunk 1]
```diff
@@ -210,6 +222,34 @@ def dmean(x_predict: np.ndarray, x_train: np.ndarray, kern: GPy.kern, w_vec: np.
...
+        where res[:, :, i, j] is the gradient of the covariance between outputs at x1 and x2 with respect to x1[i, j]
+    """
+    q1, q2, d, n = x1.shape[0], x2.shape[0], x1.shape[1], x_train.shape[0]
+    dkx1X_dx = np.zeros((d, q1*q1, n))
+    dkx1x2_dx = np.zeros((d, q1*q1, q2))
+    for i in range(d):
+        dkx1X_dx[i, ::q1 + 1, :] = kern.dK_dX(x1, x_train, i)
+        dkx1x2_dx[i, ::q1 + 1, :] = kern.dK_dX(x1, x2, i)
+    dkx1X_dx = dkx1X_dx.reshape((d, q1, q1, n))
+    dkx1x2_dx = dkx1x2_dx.reshape((d, q1, q1, q2))
```
Comment (apaleyes) [inline emukit/model_wrappers/gpy_model_wrappers.py:246]: Naming here is hard to get through: `dkx1X`? Can it be improved?

File: emukit/model_wrappers/gpy_model_wrappers.py [hunk 2]
```diff
@@ -61,6 +61,19 @@ def get_joint_prediction_gradients(self, X: np.ndarray) -> Tuple[np.ndarray, np.
...
+        """
+        Computes and returns model gradients of the covariance between outputs at points X1 and X2 with respect
+        to X1.
+
+        :param X1: points to compute gradients at, nd array of shape (q1, d)
+        :param X2: points for the covariance of which to compute the gradient, nd array of shape (q2, d)
+        :return: gradient of the covariance matrix of shape (q1, q2) between outputs at X1 and X2
+                 (return shape is (q1, q2, q1, d)).
+        """
+        dcov_dx1 = dCov(X1, X2, self.model.X, self.model.kern, self.model.posterior.woodbury_inv)
```
Comment (apaleyes) [inline emukit/model_wrappers/gpy_model_wrappers.py:74]: Three inputs come directly from the model. I wander if it makes sense to just pass model in. Or not extract `dCov` as a stateless method at all. COnisder this: all this method is doing is just calling dCov, literally nothing else. And so far dCov is only used here. It seems pretty specific to me to not get reused. Is that impression correct?

**Review by apaleyes — COMMENTED** [review]
I have two major concerns:

1. This PR introduces a new method that isn't declared in any model interfaces. That's backwards from Emukit's point of view. Are you planning to use this method anywhere? Perhaps create a new acquisition? If so, let's define an interface, or augment an existing one if that makes more sense.
2. Both changes carry some pretty heavy matrix algebra, which is really hard to follow. Can you please add some comments throughout, to clarify what's going on? Thanks!

There are also some minor comments, which can be found below.

File: emukit/model_wrappers/gpy_model_wrappers.py (hunk 2 above)
Comment (BrunoKM) [reply emukit/model_wrappers/gpy_model_wrappers.py:74]: I think that's fair to say, I was primarily just trying to mirror the implementation of `dSigma` that was already in the module. Happy to not separate it out.

**Comment by BrunoKM** [comment]
To answer the two points:

> 1. This PR introduces a new method that isn't declared in any model interfaces. That's backwards from Emukit's point of view. Are you planning to use this method anywhere? Perhaps create a new acquisition? If so, let's define an interface, or augment an existing one if that makes more sense.
This is indeed intended for an implementation of a new acquisition. It comes in handy whenever optimizing any of many batch acquisition functions sequentially – i.e. by selecting optimising batch acquisition with respect to one point in the batch at a time.  Currently, `get_covariance_between_points` is part of the `IEntropySearchModel` interface, so maybe something like `IDifferentiableEntropySeachModel` would make sense?
> 2. Both changes carry some pretty heavy matrix algebra, which is really hard to follow. Can you please add some comments throughout, to clarify what's going on? Thanks!
Fair point, I'll try and make it easier to follow.

File: emukit/model_wrappers/gpy_model_wrappers.py [hunk 3]
```diff
@@ -118,6 +118,40 @@ def get_covariance_between_points(self, X1: np.ndarray, X2: np.ndarray) -> np.nd
...
+    def get_covariance_between_points_gradients(self, X1: np.ndarray, X2: np.ndarray) -> np.ndarray:
+        """
+        Compute the derivative of the posterior covariance matrix between prediction at inputs x1 and x2
+        with respect to x1.
+
+        :param x1: Prediction inputs of shape (q1, d)
+        :param x2: Prediction inputs of shape (q2, d)
+        :param x_train: Training inputs of shape (n_train, d)
+        :param kern: Covariance of the GP model
+        :param w_inv: Woodbury inverse of the posterior fit of the GP
```
Comment (apaleyes) [inline emukit/model_wrappers/gpy_model_wrappers.py:130]: I suppose this is a bit out of date now?

File: emukit/model_wrappers/gpy_model_wrappers.py [hunk 4]
```diff
@@ -118,6 +118,40 @@ def get_covariance_between_points(self, X1: np.ndarray, X2: np.ndarray) -> np.nd
...
+        :return: nd array of shape (q1, q2, d) representing the gradient of the posterior covariance between x1 and x2
+            with respect to x1. res[i, j, k] is the gradient of Cov(y1[i], y2[j]) with respect to x1[i, k]
+        """
+        # Get the relevent shapes
```
Comment (apaleyes) [inline emukit/model_wrappers/gpy_model_wrappers.py:134]:
```suggestion
        # Get the relevant shapes
```

File: emukit/model_wrappers/gpy_model_wrappers.py [hunk 5]
```diff
@@ -118,6 +118,40 @@ def get_covariance_between_points(self, X1: np.ndarray, X2: np.ndarray) -> np.nd
...
+        q1, q2, input_dim, n_train = X1.shape[0], X2.shape[0], X1.shape[1], self.model.X.shape[0]
+        # Instatiate an array to hold gradients of prior covariance between outputs at X1 and X_train
+        cov_X1_Xtrain_grad = np.zeros((input_dim, q1, n_train))
+        # Instatiate an array to hold gradients of prior covariance between outputs at X1 and X2
```
Comment (apaleyes) [inline emukit/model_wrappers/gpy_model_wrappers.py:138]:
```suggestion
        # Instantiate an array to hold gradients of prior covariance between outputs at X1 and X2
```

File: emukit/model_wrappers/gpy_model_wrappers.py [hunk 6]
```diff
@@ -182,24 +216,23 @@ def dSigma(x_predict: np.ndarray, x_train: np.ndarray, kern: GPy.kern, w_inv: np
...
     """
     q, d, n = x_predict.shape[0], x_predict.shape[1], x_train.shape[0]
-    dkxX_dx = np.empty((q, n, d))
-    dkxx_dx = np.empty((q, q, d))
+    # Tensor for the gradients of (q, n) covariance matrix between x_predict and x_train with respect to
+    # x_predict (of shape (q, d))
+    dkxX_dx = np.zeros((d, q*q, n))
+    # Tensor for the gradients of full covariance matrix at points x_predict (of shape (q, q) with respect to
+    # x_predict (of shape (q, d))
+    dkxx_dx = np.zeros((d, q*q, q))
```
Comment (apaleyes) [inline emukit/model_wrappers/gpy_model_wrappers.py:224]: Naming here is a bit unfortunate. I know it was already there, but do you have any ideas how to improve it? the only way to tell the two vars apart is one upper/lower case X, and that's so poor to read

**Review by apaleyes — COMMENTED** [review]
I think we are good to go here, just one outdated docstring to update

**Comment by mmahsereci** [comment]
What about the interface discussion? Is this resolved? Where exactly is the new method required in Emukit, i.e., is there an acquisition function that requires the method?

**Comment by apaleyes** [comment]
Hey @mmahsereci , good to hear from you again!

@BrunoKM promised us an acquisition later down the line. But that's a good point that I forgot all about - the function shall be there as an implementation of some interface. Thanks for catching!

**Comment by mmahsereci** [comment]
That sounds great, thanks for catching me up @apaleyes! And, yes I am back now it seems. :)

**Comment by BrunoKM** [comment]
@apaleyes sorry, resolving the interfaces has been on the stack of my todos for the past couple of weeks, I will get around to it soon.

Yes, this is for a new acquisition that is hopefully still in the pipeline :)

**Comment by apaleyes** [comment]
@BrunoKM are you still planning to work on this PR (and the follow up of it)? Or shall we just close it?

**Comment by BrunoKM** [comment]
Thanks for the follow-up @apaleyes!

I'm still happy to finish off this PR (if I don't do it before the end-of-year, feel free to close it).

I've been a bit on the fence about whether to push the new acquisition soon afterwards. It's one with particularly tricky gradients: i.e. manually writing a gradient function for it in pure NumPy turned out to be a pain. The current implementation internally uses either jax/pytorch to automatically get the gradient function. I'd assume adding these dependencies is not in line with the philosophy of the rest of the EmuKit repo.

I'm still hoping I'll be able to add the manual gradients later down the line and push the new acquisition function, but due to limited time, I can't say how soon that might be.

File: emukit/model_wrappers/gpy_model_wrappers.py (hunk 3 above)
Comment (BrunoKM) [reply emukit/model_wrappers/gpy_model_wrappers.py:130]: Removed redundant flags

File: emukit/model_wrappers/gpy_model_wrappers.py (hunk 4 above)
Comment (BrunoKM) [reply emukit/model_wrappers/gpy_model_wrappers.py:134]: Fixed

File: emukit/model_wrappers/gpy_model_wrappers.py (hunk 6 above)
Comment (BrunoKM) [reply emukit/model_wrappers/gpy_model_wrappers.py:220]: I renamed it to something more verbose, let me know if you prefer it!

Comment (BrunoKM) [reply emukit/model_wrappers/gpy_model_wrappers.py:220]: I tried to improve it, let me know what you think!

**Comment by BrunoKM** [comment]
I added an interface that specifies the covariance between points derivative method. Let me know your thoughts!

As for the new acquisition function, I think I'll be able to add it without an `evaluate_with_gradients` method relatively soon. An update with `evaluate_with_gradients` could follow later down the line.

File: emukit/core/interfaces/models.py [hunk 7]
```diff
@@ -72,6 +72,35 @@ def get_joint_prediction_gradients(self, X: np.ndarray) -> Tuple[np.ndarray, np.
         raise NotImplementedError
 
 
+class ICrossCovarianceDifferentiable:
+    def get_covariance_between_points(self, X1: np.ndarray, X2: np.ndarray) -> np.ndarray:
```
Comment (apaleyes) [inline emukit/core/interfaces/models.py:76]: We already have an interface that defines nearly the same method: https://github.com/EmuKit/emukit/blob/main/emukit/bayesian_optimization/interfaces/models.py

While this isn't that big of an issue, would be nice if we could separate them somehow. Few ideas:
* Just rename this method
* Rename the other mehtod
* Drag this method into a separate interface

Pick whichever you prefer!

**Review by apaleyes — COMMENTED** [review]
looks great, just one comment

**Comment by apaleyes** [comment]
Unfortunately I think we shall be closing this PR. It was a prep for a new acq function, so it does not make sense to push it on its own, without a follow up.
